search/search_quality/samples_generation_tool/samples_generation_tool.cpp

```diff
@@ -0,0 +1,458 @@
+#include "search/categories_cache.hpp"
+#include "search/mwm_context.hpp"
+#include "search/reverse_geocoder.hpp"
+#include "search/search_quality/helpers.hpp"
+#include "search/search_quality/sample.hpp"
+
+#include "indexer/classificator_loader.hpp"
+#include "indexer/data_source.hpp"
+#include "indexer/feature_algo.hpp"
+#include "indexer/ftypes_matcher.hpp"
+#include "indexer/search_string_utils.hpp"
+
+#include "coding/file_name_utils.hpp"
+
+#include "platform/platform.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/macros.hpp"
+#include "base/string_utils.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <ctime>
+#include <fstream>
+#include <limits>
+#include <map>
+#include <random>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "3party/gflags/src/gflags/gflags.h"
+
+using namespace search::search_quality;
+using namespace search;
+using namespace std;
+
+DEFINE_string(data_path, "", "Path to data directory (resources dir)");
+DEFINE_string(mwm_path, "", "Path to mwm files (writable dir)");
+DEFINE_string(out_buildings_path, "buildings.json", "Path to output file for buildings samples");
+DEFINE_string(out_cafes_path, "cafes.json", "Path to output file for cafes samples");
+
+double constexpr kMaxDistanceToObjectM = 7500.0;
+double constexpr kMinViewportSizeM = 100.0;
+double constexpr kMaxViewportSizeM = 5000.0;
+size_t constexpr kMaxSamplesPerMwm = 20;
+
+std::random_device rd;
+mt19937 rng(rd());
+
+enum class RequestType
+{
+  Building,
+  Cafe
+};
+
+bool GetRandomBool()
+{
+  std::uniform_int_distribution<> dis(0, 1);
+  return dis(rng) == 0;
+}
+
+// Leaves first letter as is.
+void EraseRandom(strings::UniString & uni)
+{
+  if (uni.size() <= 1)
+    return;
+
+  uniform_int_distribution<size_t> dis(1, uni.size() - 1);
+  auto const it = uni.begin() + dis(rng);
+  uni.erase(it, it + 1);
+}
+
+// Leaves first letter as is.
+void SwapRandom(strings::UniString & uni)
+{
+  if (uni.size() <= 2)
+    return;
+
+  uniform_int_distribution<size_t> dis(1, uni.size() - 2);
+  auto const index = dis(rng);
+  swap(uni[index], uni[index + 1]);
+}
+
+void AddRandomMisprint(strings::UniString & str)
+{
+  // todo(@t.yan): Disable for hieroglyphs, consider implementing InsertRandom.
+  if (GetRandomBool())
+    return EraseRandom(str);
+  return SwapRandom(str);
+}
+
+void AddMisprints(string & str)
+{
+  auto tokens = strings::Tokenize(str, " -&");
+  str.clear();
+  for (size_t i = 0; i < tokens.size(); ++i)
+  {
+    auto const & token = tokens[i];
+    auto uni = strings::MakeUniString(token);
+    if (uni.size() > 4 && GetRandomBool() && GetRandomBool())
```

Comment:
I think ```rng() % 4 == 0``` is more readable and enough for our needs (and taking mod by small powers of two also seems safe)

Reply:
done